### PR TITLE
try to fix coordinates

### PR DESCRIPTION
**Pull request by blankdots**
does this fix the coordinates ... well yes :) does it remove a lot of the unit tests ... yes
Tests will be revamped when we move to new repo with issue https://github.com/neicnordic/sensitive-data-archive/issues/410

inspired by: https://github.com/neicnordic/crypt4gh/blob/master/examples/reader/main.go

- disabled logging health endpoint as it generates unnecessary logs

### CODE REVIEW

**Review by teemukataja — APPROVED** [review]

File: api/sda/sda.go [hunk 1]
```diff
@@ -206,90 +196,104 @@ func Download(c *gin.Context) {
...
+	log.Debug("begin data stream")
+
+	if start != 0 {
+		// We don't want to read from start, skip ahead to where we should be
+		if _, err := c4ghr.Seek(start, 0); err != nil {
+			w.WriteHeader(500)
+			_, err := w.Write([]byte(fmt.Sprintf("error occurred %v", err)))
+			if err != nil {
+				return
+			}
```
Comment (pontus) [inline api/sda/sda.go:253]:
```suggestion
			w.Write([]byte(fmt.Sprintf("error occurred %v", err)))
```

We're doing a valueless return afterwards anyway (within the seek failed branch).

(Not tested, could possibly have linter opinions causing an explicit ignore to be better.)

Comment (blankdots) [reply api/sda/sda.go:253]: `api/sda/sda.go:250:11: Error return value of `w.Write` is not checked (errcheck)` 

i will return an error there

**Comment by pontus** [comment]
Code looks good, but I'm a bit worried about the amount of unit tests cut, do all need to go? I see the added other tests ("integration"), but think we could have both.

**Comment by pontus** [comment]
Also, I'm curious, I hope (and believe) I didn't break anything when I added random access to the crypt4gh golang.

**Review by pontus — APPROVED** [review]

**Comment by blankdots** [comment]
> Also, I'm curious, I hope (and believe) I didn't break anything when I added random access to the crypt4gh golang.

We were using the datalistpackets, and I am sure they worked previously. I remember testing, but we don't have a test to see when and if they were actually broken

**Comment by blankdots** [comment]
marking as draft as we test more things

**Comment by blankdots** [comment]
@pontus we tried to make it work with large file chunks but we see that chunks larged than ~5Mb don't work
we pinpointed to: https://github.com/neicnordic/crypt4gh/pull/83 changes as after those changes the way we did it previously worked

**Comment by blankdots** [comment]
@pontus i made it work with the example reader you created; it was my mistake trying to be fancy about how to work with it, now sda-download works with larger files as well, however the problem with the data edit list still exists in crypt4gh how we were using it previously

**Review by teemukataja — APPROVED** [review]

File: api/sda/sda.go [hunk 2]
```diff
@@ -206,90 +195,106 @@ func Download(c *gin.Context) {
...
+		return
+	}
 }
 
-// sendStream streams file contents from a reader
-var sendStream = func(w http.ResponseWriter, file io.Reader) {
-	log.Debug("begin data stream")
+// sendStream
+// used from: https://github.com/neicnordic/crypt4gh/blob/master/examples/reader/main.go#L48C1-L113C1
+var sendStream = func(reader *streaming.Crypt4GHReader, writer http.ResponseWriter, start, end int64) error {
```
Comment (pontus) [inline api/sda/sda.go:257]: Could this possibly be simplified by using `reader.Seek` at start and then use of `io.CopyN`? (Will still need to loop I think, but could be a bit nicer).

Comment (blankdots) [reply api/sda/sda.go:257]: i will try to improve that when switching to `neicnordic/sensitive-data-archive`, till then I will leave as is https://github.com/neicnordic/sensitive-data-archive/issues/410

**Comment by blankdots** [comment]
> When this PR is merged can we archive this repo so that all future work gets done in [neicnordic/sensitive-data-archive](https://github.com/neicnordic/sensitive-data-archive)?
> Also these changes needs to be moved into the `sensitive-data-archive` repo as well.

yes, but there were some issues @aaperis reported i will see if i can replicate

File: .github/integration/tests/s3notls/52_check_endpoint.sh [hunk 3]
```diff
@@ -86,6 +86,30 @@ else
...
+curl -H "Authorization: Bearer $token" "http://localhost:8080/files/urn:neic:001-002?startCoordinate=7&endCoordinate=14" --output test-part2.txt
+
+echo "LE IS J" > old-part2.txt
+
+cmp --silent old-part2.txt test-part2.txt
+status=$?
+if [[ $status = 0 ]]; then
+    echo "Files are the same"
+else
+    echo "Files are different"
```
Comment (aaperis) [inline .github/integration/tests/s3notls/52_check_endpoint.sh:110]:
```suggestion
    echo "Files are different"
    exit 1
```

File: .github/integration/tests/s3notls/52_check_endpoint.sh [hunk 4]
```diff
@@ -86,6 +86,30 @@ else
...
+curl -H "Authorization: Bearer $token" "http://localhost:8080/files/urn:neic:001-002?startCoordinate=0&endCoordinate=2" --output test-part.txt
+
+echo "TH" > old-part.txt
+
+cmp --silent old-part.txt test-part.txt
+status=$?
+if [[ $status = 0 ]]; then
+    echo "Files are the same"
+else
+    echo "Files are different"
```
Comment (aaperis) [inline .github/integration/tests/s3notls/52_check_endpoint.sh:98]:
```suggestion
    echo "Files are different"
    exit 1
```

File: .github/integration/tests/common/50_check_endpoint.sh [hunk 5]
```diff
@@ -90,6 +90,30 @@ else
     echo "Files are different"
 fi
 
+curl --cacert certs/ca.pem -H "Authorization: Bearer $token" "https://localhost:8443/files/urn:neic:001-002?startCoordinate=0&endCoordinate=2" --output test-part.txt
+
+echo "TH" > old-part.txt
```
Comment (aaperis) [inline .github/integration/tests/common/50_check_endpoint.sh:95]:
```suggestion
dd if=old-file.txt ibs=1 skip=0 count=2 > old-part.txt
```

File: .github/integration/tests/common/50_check_endpoint.sh [hunk 6]
```diff
@@ -90,6 +90,30 @@ else
...
+status=$?
+if [[ $status = 0 ]]; then
+    echo "Files are the same"
+else
+    echo "Files are different"
+fi
+
+curl --cacert certs/ca.pem -H "Authorization: Bearer $token" "https://localhost:8443/files/urn:neic:001-002?startCoordinate=7&endCoordinate=14" --output test-part2.txt
+
+echo "LE IS J" > old-part2.txt
```
Comment (aaperis) [inline .github/integration/tests/common/50_check_endpoint.sh:107]:
```suggestion
dd if=old-file.txt ibs=1 skip=7 count=7 > old-part2.txt
```

File: .github/integration/tests/s3notls/52_check_endpoint.sh [hunk 7]
```diff
@@ -86,6 +86,30 @@ else
     echo "Files are different"
 fi
 
+curl -H "Authorization: Bearer $token" "http://localhost:8080/files/urn:neic:001-002?startCoordinate=0&endCoordinate=2" --output test-part.txt
+
+echo "TH" > old-part.txt
```
Comment (aaperis) [inline .github/integration/tests/s3notls/52_check_endpoint.sh:91]:
```suggestion
dd if=old-file.txt ibs=1 skip=0 count=2 > old-part.txt
```

File: .github/integration/tests/s3notls/52_check_endpoint.sh [hunk 8]
```diff
@@ -86,6 +86,30 @@ else
...
+status=$?
+if [[ $status = 0 ]]; then
+    echo "Files are the same"
+else
+    echo "Files are different"
+fi
+
+curl -H "Authorization: Bearer $token" "http://localhost:8080/files/urn:neic:001-002?startCoordinate=7&endCoordinate=14" --output test-part2.txt
+
+echo "LE IS J" > old-part2.txt
```
Comment (aaperis) [inline .github/integration/tests/s3notls/52_check_endpoint.sh:103]:
```suggestion
dd if=old-file.txt ibs=1 skip=7 count=7 > old-part2.txt
```

**Review by aaperis — CHANGES_REQUESTED** [review]
I believe the problem was that the logic for seeking the first byte where the byte range will be read was missing. As a consequence, `download` was always starting from byte 1.  With the suggested change, `sda-download`, the read example from the `crypt4gh` repo and `dd` tests all agree. I also made some suggestions about the integration tests. Ideally, for the later we should use a small bam file when we migrate to the merged repo.

File: api/sda/sda.go [hunk 9]
```diff
@@ -206,90 +195,106 @@ func Download(c *gin.Context) {
...
 
-	n, err := io.Copy(w, file)
-	log.Debug("end data stream")
+	// Calculate how much we should read (if given)
```
[the remark this answers is not in the document]
Comment (blankdots) [reply api/sda/sda.go:258]: oh no :disappointed: my bad

**Comment by blankdots** [comment]
> I believe the issue was that the logic for seeking the first byte where the byte range will be read was missing.

you are right, i think i had it there, but in the confusion of me rewriting that piece many times and bad tests, i did not spot it

**Review by aaperis — APPROVED** [review]

**Pull request merged**